src/system/SystemPacketBuffer.h

```diff
@@ -230,3 +243,87 @@ inline uint16_t PacketBuffer::AllocSize() const
 
 } // namespace System
 } // namespace chip
+
+namespace chip {
+namespace System {
+
+/// Tracks ownership of a System::PacketBuffer.
+class DLL_EXPORT PacketBufferHandle
+{
+public:
+    PacketBufferHandle() : mBuffer(nullptr) {}
+    PacketBufferHandle(decltype(nullptr)) : mBuffer(nullptr) {}
+
+    // The caller's ownership is transferred to this.
+    explicit PacketBufferHandle(PacketBuffer * buffer) : mBuffer(buffer) {}
+
+    PacketBufferHandle(PacketBufferHandle && aOther)
+    {
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+    }
+
+    ~PacketBufferHandle() { Adopt(nullptr); }
+    void Free() { Adopt(nullptr); }
+
+    PacketBufferHandle & operator=(PacketBufferHandle && aOther)
+    {
+        if (mBuffer)
+        {
+            PacketBuffer::Free(mBuffer);
+        }
+        mBuffer        = aOther.mBuffer;
+        aOther.mBuffer = nullptr;
+        return *this;
+    }
+    PacketBufferHandle & operator=(decltype(nullptr))
+    {
+        Adopt(nullptr);
+        return *this;
+    }
+
+    PacketBufferHandle Clone() const
+    {
+        mBuffer->AddRef();
+        return PacketBufferHandle(mBuffer);
+    }
+
+    PacketBuffer * operator->() { return mBuffer; }
+    PacketBuffer & operator*() { return *mBuffer; }
+
+    const PacketBuffer * operator->() const { return mBuffer; }
+    const PacketBuffer & operator*() const { return *mBuffer; }
+
+    // The caller's ownership is transferred to this.
+    void Adopt(PacketBuffer * buffer)
+    {
+        if (mBuffer != nullptr)
+        {
+            PacketBuffer::Free(mBuffer);
+        }
+        mBuffer = buffer;
+    }
+
+    // The PacketBufferHandle's ownership is transferred to the caller.
+    CHECK_RETURN_VALUE PacketBuffer * Release()
+    {
+        PacketBuffer * buffer = mBuffer;
+        mBuffer               = nullptr;
+        return buffer;
+    }
+
+    // The caller has access but no ownership.
+    // This is intended to be used only to call functions that have not yet been converted to take a PacketBufferHandle.
+    [[deprecated]] PacketBuffer * Get_NoRelease() { return mBuffer; }
```

Comment:
This just a nuisance for developers and early integrators - CHIP triggers **its own diagnostics** unless building with Wno-deprecated-declarations.

Which we do upstream, because we are misusing this specifier. Please don't.

Reply:
Done